nixos/modules/system/boot/luksroot.nix

```diff
@@ -221,33 +241,13 @@ let
 
     # LUKS
     open_normally() {
-        ${if (dev.keyFile != null) then ''
-        if wait_target "key file" ${dev.keyFile}; then
-            ${csopen} --key-file=${dev.keyFile} \
-              ${optionalString (dev.keyFileSize != null) "--keyfile-size=${toString dev.keyFileSize}"} \
-              ${optionalString (dev.keyFileOffset != null) "--keyfile-offset=${toString dev.keyFileOffset}"}
-            cs_status=$?
-            if [ $cs_status -ne 0 ]; then
-              echo "Key File ${dev.keyFile} failed!"
-              if ! try_empty_passphrase; then
-                ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
-                echo " - failing back to interactive password prompt"
-                do_open_passphrase
-              fi
-            fi
-        else
-            # If the key file never shows up we should also try the empty passphrase
-            if ! try_empty_passphrase; then
-               ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
-               echo " - failing back to interactive password prompt"
-               do_open_passphrase
-            fi
+        '' + keyAttempts + ''
+        if ! try_empty_passphrase; then
+            ${if dev.fallbackToPassword then
+                 ''echo "Falling back to passphrase"''
+              else ''die "Failed to unlock ${dev.name}"''}
+            do_open_passphrase
         fi
-        '' else ''
-           if ! try_empty_passphrase; then
-              do_open_passphrase
-           fi
-        ''}
```

Comment:
Tried to simplify the logic a bit. Hard to unify all code paths with existing options